Feature hcc opt flush2

bensander opened the pull request:
After cleaning up indentation nightmare.
Passes HIP tests as well, including with HCC_MAX_QUEUES=2.

Default results from HCC tests below:

********************
Testing Time: 702.16s
********************
Unexpected Passing Tests (4):
    HCC :: Unit/FilePath/file path_test2.cpp
    HCC :: Unit/FilePath/file_path_test1.cpp
    HCC :: Unit/FilePath/file_path_test3.cpp
    HCC :: Unit/FilePath/file_path_test4.cpp

********************
Failing Tests (3):
    HCC :: Unit/HSA/list2.cpp
    HCC :: benchEmptyKernel/bench.cpp
    HCC :: benchEmptyKernel/nullkernel.cpp

  Expected Passes    : 691
  Expected Failures  : 21
  Unexpected Passes  : 4
  Unexpected Failures: 3
make[3]: *** [tests/CMakeFiles/test] Error 1
make[2]: *** [tests/CMakeFiles/test.dir/all] Error 2
make[1]: *** [tests/CMakeFiles/test.dir/rule] Error 2

bensander committed:
Refactor HCC environment variable interface.

Define common function which ensures variable name matches env string.
Env var specified once.
Add HCC_PRINT_ENV which prints env var state when app runs.

bensander committed:
Fix accelerator_view::set_cu_mask.

Has been broken for a while, the interface to hsa_amd_queue_cu_set_mask
requires size be a multiple of 32.
accelerator_view_set_cu_mask test now passes.

bensander committed:
Support cu_mark with HCC_MAX_QUEUES.

Overall strategy is to :
- save the desired CU mask with the accelerator_view structure,
- save actual cu_mask set on the current hw hsa_queue in RocrQueue.

When queues are stolen, we check the desired state against the hw state
and call hsa_amd_queue_cu_set_mask if they differ.  this is an
expensive operation so scheme will not perform well if cu_mask
is combined with queue oversubscription.

Conflicts:
	lib/hsa/mcwamp_hsa.cpp

bensander committed:
Fix bug in steal queues.

Also move env var reading to GetHccVar free function.

Conflicts:
	lib/hsa/mcwamp_hsa.cpp

bensander committed:
fix get_all_queues scrambled during merge

bensander committed:
Merge branch 'feature_hcc_opt_flush' into clang_tot_upgrade

Conflicts:
	clang
	lib/hsa/mcwamp_hsa.cpp

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -39,6 +39,7 @@
 #include <time.h>
 #include <iomanip>
 
+#define KALMAR_DEBUG (0)
```
this line seems redundant?

bensander replied:
This should be removed.  We just need the one inside the ifdef.

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -763,11 +767,56 @@ pool_iterator::pool_iterator()
+    ~RocrQueue() {
+
+#if KALMAR_DEBUG
+        std::cerr << "HSAQueue::dispose(): destroy an HSA command queue: " << _hwQueue << "\n";
```
should be better changed to `RocrQueue::~RocrQueue()` to avoid confusion

bensander replied:
This is inside the structure definition so can't use full namespace path.

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -781,11 +830,15 @@ class HSAQueue final : public KalmarQueue
 
     uint64_t                                      opSeqNums;
 
+    bool                                            valid;
```
@bensander could you provide a comment on the meaning of `valid` here? seems it's being used in quite a few places.

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -1021,7 +1008,7 @@ class HSAQueue final : public KalmarQueue
         int count = 0;
         for (int i = 0; i < asyncOps.size(); ++i) {
-            auto asyncOp = asyncOps[i];
+            auto asyncOp = asyncOps[i]; // TODO _ shared_ptr copy, can we optimize?
```
would `auto&` help reduce the copy?

pfultz2 replied:
Actually, since you don't use the index, you could just write:

```cpp
for(auto&& asyncOp:asyncOps) {
    if (asyncOp != nullptr) {
        hsa_signal_t signal = *(static_cast <hsa_signal_t*> (asyncOp->getNativeHandle()));
        hsa_signal_value_t v = hsa_signal_load_relaxed(signal);
        if (v != 0) {
            ++count;
        }
    }
}
```

pfultz2 replied:
Or even better:

```cpp
return std::count_if(asyncOps.begin(), asyncOps.end(), [](auto&& asyncOp)
{
    return asyncOp != nullptr and 
        hsa_signal_load_relaxed(*(static_cast <hsa_signal_t*> (asyncOp->getNativeHandle()))) != 0;
});
```

bensander replied:
I change to reference.   The index i is useful for gdb debug at times.

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -1034,6 +1021,24 @@ class HSAQueue final : public KalmarQueue
+        // Have to walk asyncOps since it can contain null pointers.
+        for (int i = 0; i < asyncOps.size(); ++i) {
+            if (asyncOps[i] != nullptr) {
+                auto asyncOp = asyncOps[i];
```
would `auto&` reduce the copy?

pfultz2 replied:
Likewise this could be written as:

```cpp
return not std::any_of(asyncOps.begin(), asyncOps.end(), [](auto&& asyncOp)
{
    return asyncOp != nullptr and 
        hsa_signal_load_relaxed(*(static_cast <hsa_signal_t*> (asyncOp->getNativeHandle()))) != 0;
});
```

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -1607,6 +1673,106 @@ class HSADevice final : public KalmarDevice
+            status = hsa_amd_profiling_set_profiler_enabled(hwQueue, 1);
+
+
+            auto rq = new RocrQueue(hwQueue, thief);
```
@bensander could you explain the logic here a bit? It seems you are creating 2 `RocrQueue` instances, `rq` and `rocrQueue`, and push them into `rocrQueues`. Is this really your intended behavior?

bensander replied:
Looks like the merge duplicated some code, I will clean it up in next push.

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -1607,6 +1673,106 @@ class HSADevice final : public KalmarDevice
+            
+
+        } else {
+            while (1) {
```
could we use a `bool` flag? I feel a bit uneasy to see `while(1)`...

scchan replied:
Just wondering if we need to release and then re-acquire the mutex for rocrQueuesMutex after each iteration to give a chance for other threads to release their queue resource while the current thread is spinning looking for an empty queue>

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -1607,6 +1673,106 @@ class HSADevice final : public KalmarDevice
+                        // update the queue pointers to indicate the theft:
+                        rq->assignHccQueue(thief);
+
+                        return;
```
and change the flag to break the while loop.

whchung commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -763,11 +767,56 @@ pool_iterator::pool_iterator()
+    friend class RocrQueue;
+
+    // ROCR queue associated with this HSAQueue instance. 
+    RocrQueue    *rocrQueue;
```
@bensander could you add a bit more comment on the relationship between `HSAQueue` and `RocrQueue`? It seems now `RocrQueue` holds the real handle returned by `hsa_queue_create()`, and `HSAQueue` is a logical instance, and multiple `HSAQueue` instances may share one `RocrQueue`. Is this understanding correct?

bensander replied:
Yes.  
Really I think we should rename HSAQueue to something like HSAAcceleratorView - this is actually the implementation of accelerator_view for the "hsa" backend.  
Then RocrQueue could be HSAQueue, this is thin wrapper around the hsa_queue_t.

whchung reviewed (changes requested):
@bensander I've finished my first round of review. could you help check my questions? thanks.

pfultz2 commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -1034,6 +1021,24 @@ class HSAQueue final : public KalmarQueue
 
+
+    int isEmpty() override {
+        int count = 0;
```
Count is not used here.

scchan commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -763,11 +767,56 @@ pool_iterator::pool_iterator()
+    RocrQueue(hsa_queue_t *hwQueue, HSAQueue *hccQueue) : 
+        _hwQueue(hwQueue), _hccQueue(hccQueue) 
+    {
+        assignHccQueue(hccQueue);
```
Can we move the hsa_queue creation into the RocrQueue's constructor rather than passing it in as an argument?  I think this is more consistent with the ~RocrQueue where the hsa_queue gets destroyed.

bensander replied:
Good idea.

bensander committed:
Refactor to incorporate review feedback:

- Move queue creation inside RocrQueue contructor.
- Fix duplicated RocrQ creation from merge.
- Add comments explaining RocrQueue vs HSAQueue, valid flag.
- Refactor RocrQ acquisition loop to use flag, release queues lock in
  each loop.
- Use references in asyncOps loop to avoid unneeded shared ptr copy.

bensander committed:
Fix keep-around modei for rcor-queue dealloc.  Enhance HCC_DB.

When removing an HCC queue, only remove the HW RocrQ if the hccQueues <
rocrQueues.    This allows other hccQueues in the future to grab an
available rocrQueue from pool without having to create a new one.
But, if no other hccQueues exist, then shrink the pool.

bensander commented:
Folks please review updated code.

scchan commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -763,11 +771,83 @@ pool_iterator::pool_iterator()
+
+        assert (__builtin_popcount(MAX_INFLIGHT_COMMANDS_PER_QUEUE) == 1); // make sure this is power of 2.
+        assert(queue_size > MAX_INFLIGHT_COMMANDS_PER_QUEUE*2);
+
```
Maybe we should move these asserts into HSADevice such that we would only perform them once per device rather than for every queue creation?

scchan commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -1439,44 +1489,69 @@ class HSAQueue final : public KalmarQueue
         // enqueue the barrier
-        status = barrier.get()->enqueueAsync(this);
+        status = barrier.get()->enqueueAsync(this, false/*agent-flush*/);
+        STATUS_CHECK(status, __LINE__);
```
I think this is changing the semantics of EnqueueMarker (and also EnqueueMarkerWithDependency) by changing it from system scope to agent scope fence.  I think we expose these through the hc::create_marker API so this will break user code the marker was used as the synchronization mechanism between the host and the device?

bensander replied:
The logic is a little confusing but this second parm is "forceFlush" and only force the fences to agent scope if HCC_OPT_FLUSH=1 - which is not enabled yet.

    unsigned fenceBits = (!forceFlush && HCC_OPT_FLUSH) ?
        ((HSA_FENCE_SCOPE_AGENT) << HSA_PACKET_HEADER_ACQUIRE_FENCE_SCOPE) |
        ((HSA_FENCE_SCOPE_AGENT) << HSA_PACKET_HEADER_RELEASE_FENCE_SCOPE)
        :
        ((HSA_FENCE_SCOPE_SYSTEM) << HSA_PACKET_HEADER_ACQUIRE_FENCE_SCOPE) |
        ((HSA_FENCE_SCOPE_SYSTEM) << HSA_PACKET_HEADER_RELEASE_FENCE_SCOPE);

Eventual behavior will be that Markers have System scope by default, but a flag to create_marker can select agent scope in cases where this is all that matters.

bensander committed:
Move queue size computation&check so it is done once/device.

Also disable KALMAR_DEBUG flag by default.

scchan commented:
Approved, thanks!
